File: projects/spring-boot-restv2/demo/src/main/java/com/sourabh/demo/service/RoomService.java
package com.sourabh.demo.service;

import com.sourabh.demo.exceptions.SomethingNotFoundException;
import com.sourabh.demo.iservice.IRoomService;
import com.sourabh.demo.model.Room;
import com.sourabh.demo.repository.RoomRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;


import java.util.List;


@Service
public class RoomService implements IRoomService {

    @Autowired
    RoomRepository roomRepository;


    public Room findById(Integer id) {
        return roomRepository.findById(id).orElseThrow(() -> new SomethingNotFoundException("Room not found for the id: " + id));
    }
    public Room findByIdAvailable(Integer id) {
        Room room= roomRepository.findById(id).orElseThrow(() -> new SomethingNotFoundException("Room not found for the id: " + id));
        if(room.isAvailable())
            return room;
        else
            throw new SomethingNotFoundException("Room of specified type not available");
    }

    public Room findByType(String type) {
        Room room = roomRepository.findByType(type);
        if (room == null)
            throw new SomethingNotFoundException("No Room with type: " + type);
        return room;

    }
    public Room findByTypeAvailable(String type) {
        Room room = roomRepository.findByType(type);
        if (room == null || !room.isAvailable())
            throw new SomethingNotFoundException("No Room with type: " + type);
        return room;

    }

    public List<Room> getAll() {
        List<Room> rooms=roomRepository.findAll();
        if (rooms.isEmpty())
            throw new SomethingNotFoundException("No Rooms data available");
        return rooms;
    }
    public List<Room> getAllAvailable() {
        List<Room> rooms = roomRepository.getAllAvailable();
        if (rooms.isEmpty())
            throw new SomethingNotFoundException("NO ROOMS AVAILABLE");
        return rooms;
    }

    public Room addOrUpdate(Room room) {
        room.setAvailable(true);
        return roomRepository.save(room);
    }

    public List<Room> addAll(List<Room> rooms) {
        return roomRepository.insert(rooms);
    }

    public Room removeRoom(int id) {
        Room roomDetails = findById(id);
        roomDetails.setAvailable(false);
        return roomRepository.save(roomDetails);
    }
    public void updateRoomCount(Room room,int count){
        int updatedCount = room.getLeft()+count;
        room.setLeft(updatedCount);
        roomRepository.save(room);

    }

}

File: projects/spring-boot-rest-kt/hotel-reservation/src/main/kotlin/com/sourabh/hotelreservation/service/RoomService.kt
package com.sourabh.hotelreservation.service

import com.sourabh.hotelreservation.exception.SomethingNotFoundException
import com.sourabh.hotelreservation.iservice.IRoomService
import com.sourabh.hotelreservation.model.Room
import com.sourabh.hotelreservation.repository.RoomRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class RoomService(@Autowired private var roomRepository: RoomRepository):IRoomService {

    override fun addOrUpdate(room:Room):Room{
        room.available=true
        return roomRepository.save(room);
    }

    override fun findById(id: Int): Room {
        return roomRepository.findById(id).orElseThrow { SomethingNotFoundException("No Room found for the id: $id") }
    }

    override fun findByIdAvailable(id: Int): Room {
        val room = findById(id)
        if(room.available)
            return room
        else
            throw SomethingNotFoundException("Room with ID $id not available")
    }

    override fun findByType(type: String): Room {
        return roomRepository.findByType(type) ?: throw SomethingNotFoundException("No room with type: $type")

    }

    override fun findByTypeAvailable(type: String): Room {
        val room = roomRepository.findByType(type)
        if (room == null || !room.available)
            throw SomethingNotFoundException("No room with type: $type")
        return room
    }

    override fun getAll():MutableList<Room>?{
        val rooms=roomRepository.findAll()
        if(rooms.isEmpty())
            throw SomethingNotFoundException("No room data available")
        return rooms
    }

    override fun getAllAvailable(): MutableList<Room> {
        val rooms = roomRepository.getAllAvailable()
        if (rooms.isEmpty())
            throw SomethingNotFoundException("NO ROOMS AVAILABLE");
        return rooms
    }

    override fun addAll(rooms:MutableList<Room>): MutableList<Room> {
        return roomRepository.insert(rooms)
    }

    override fun removeRoom(id: Int): Room {
        var roomDetails = findById(id)
        roomDetails.available=false
        return roomRepository.save(roomDetails)
    }

    override fun updateRoomCount(room: Room, count:Int) {
         room.left = room.left+count
        if(room.left == 0)
            room.available=false
        roomRepository.save(room)

    }


}